adapt Apple's KVO publisher implementation

mayoff opened the pull request:
See #75. This commit adapts Apple's KVO publisher for CombineX.

Please note that Apple's implementation has a bug, which I did not fix in the adapted version. Test case 2.1 expects the buggy behavior, and has a comment describing the bug and what the correct behavior would be.

Commit:
adapt Apple's KVO publisher implementation

Commit:
suppress KeyValueObservingSpec unless Objective-C is available

The KVO publisher requires Objective-C runtime integration.

mayoff commented:
Turns out this isn't quite fully baked yet.

Commit:
move KVO publisher onto cx wrapper

mayoff commented:
Okay, take two. Now it actually puts the KVO publisher method on the cx object as it should.

ddddxxx commented on Sources/CXFoundation/Publishers+KeyValueObserving.swift:
```diff
@@ -0,0 +1,216 @@
+    /// - Returns: A publisher that emits elements each time the property’s value changes.
+    public func publisher<Value>(for keyPath: KeyPath<Base, Value>,
+                                 options: NSKeyValueObservingOptions = [.initial, .new])
+        -> CXWrappers.KeyValueObservingPublisher<Base, Value> {
```
`CXWrappers.KeyValueObservingPublisher` should be `CXWrappers.NSObject.KeyValueObservingPublisher`.

mayoff replied:
Moved.

ddddxxx commented on Tests/CXFoundationTests/KeyValueObservingSpec.swift:
```diff
@@ -0,0 +1,135 @@
+#endif
+
+class KeyValueObservingSpec: QuickSpec {
+    #if swift(>=5.1) && canImport(ObjectiveC)
```
Bypass this test for system Combine? They does't adapt KVO publisher yet.
```suggestion
    #if swift(>=5.1) && canImport(ObjectiveC) && !USE_COMBINE
```

mayoff replied:
Foundation includes the KVO publisher, I believe since Xcode 11.0. Certainly they are in Xcode 11.3.1 which is the oldest version I have handy.

mayoff replied:
These tests pass in the "combine" CI configuration.

ddddxxx replied:
That's odd... I can use it (in both Xcode 11.3.1 and 11.4.1), but [official documentation](https://developer.apple.com/documentation/objectivec/nsobject/keyvalueobservingpublisher) says `KeyValueObservingPublisher` does't conforms to `Publisher`. ⌘+Click in Xcode does't bring me to the declaration either.

ddddxxx replied:
OK CI failed elsewhere. It seems Combine changed its behavior once again.

mayoff replied:
Xcode and the documentation don't say much about the KVO publisher because it involves `_KeyValueCodingAndObservingPublishing`, which starts with an underscore. Apple doesn't document (and Xcode hides) SDK identifiers that start with `_`.

ddddxxx replied:
Yes `_KeyValueCodingAndObservingPublishing.publisher(for:options:)` is underscored and hidden. `KeyValueObservingPublisher.didChange` and conformance of `Publisher` don't involve underscore though.

I've seen similar header magic before (#29) and I still don't understand why.

ddddxxx commented on Sources/CXFoundation/Publishers+KeyValueObserving.swift:
```diff
@@ -0,0 +1,216 @@
+            }
+        }
+
+        deinit {
+//            lock.cleanupLock()
+//            downstreamLock.cleanupLock()
+        }
```
Also remove this.

mayoff replied:
I tried to change the file as little as possible so future merges from the Swift repo would be easier.

mayoff replied:
I'm willing to remove the comments if you don't care about that reason.

ddddxxx commented on Sources/CXFoundation/Publishers+KeyValueObserving.swift:
```diff
@@ -0,0 +1,216 @@
+import CombineX
+import CXUtility
+
+// The following protocol is so that we can reference `Self` in the Publisher
+// below. This is based on a trick used in the the standard library's
+// implementation of `NSObject.observe(key path)`
+//@available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+//public protocol _KeyValueCodingAndObservingPublishing {}
+
+//@available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+//extension NSObject: _KeyValueCodingAndObservingPublishing {}
+
+//@available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
```
Remove unneeded comments?

mayoff replied:
I tried to change the file as little as possible so future merges from the Swift repo would be easier.

mayoff replied:
I'm willing to remove the comments if you don't care about that reason.

ddddxxx replied:
It's OK to me.

ddddxxx reviewed (changes requested):
Thank you! This looks very good. Only a few places need improvement.

Commit:
move types into CXWrappers.NSObject

Move KeyValueObservingPublisher and KVOSubscription from CXWrappers to CXWrappers.NSObject. Note that this complicates use of KeyValueObservingPublisher because CXWrappers.NSObject is generic. To declare a value of KeyValueObservingPublisher, you have to say CXWrappers.NSObject<Base>.KeyValueObservingPublisher.
